### ISSUE TITLE
mnist_cnn.py not working out of the box

### ISSUE THREAD

**Issue by rodrigob** [fchollet/keras#617]
Running `python3 mnist_cnn.py` with latest Keras and Theano results in

```
> python3 mnist_cnn.py 
Using gpu device 0: Quadro K2000M (CNMeM is disabled)
X_train shape: (60000, 1, 28, 28)
60000 train samples
10000 test samples
Train on 60000 samples, validate on 10000 samples
Epoch 0
60000/60000 [==============================] - 290s - loss: 2.7128 - acc: 0.7091 - val_loss: 0.1638 - val_acc: 0.9501
Epoch 1
60000/60000 [==============================] - 289s - loss: 7.7192 - acc: 0.4924 - val_loss: 15.5040 - val_acc: 0.0381
Epoch 2
60000/60000 [==============================] - 288s - loss: 15.4231 - acc: 0.0431 - val_loss: 15.5040 - val_acc: 0.0381
Epoch 3
60000/60000 [==============================] - 289s - loss: 15.4266 - acc: 0.0429 - val_loss: 15.5040 - val_acc: 0.0381
Epoch 4
60000/60000 [==============================] - 289s - loss: 15.4266 - acc: 0.0429 - val_loss: 15.5040 - val_acc: 0.0381
```
To my understanding this means "not learning".

`python3 mnist_mlp.py` works fine

```
Epoch 19
2s - loss: 0.0416 - acc: 0.9861 - val_loss: 0.0677 - val_acc: 0.9815
Test score: 0.0677454245182
Test accuracy: 0.9815
```

My .theanorc looks like
```
[global]
mode = FAST_RUN
floatX = float32
device = gpu0

openmp = True

[nvcc]
fastmath = True
```

**Comment by rodrigob** [comment]
Similarly the docs of `cifar10_cnn.py` indicate "It gets down to 0.65 test logloss in 25 epochs"; yet I get

```----------------------------------------
Epoch 25
----------------------------------------
Training...
50000/50000 [==============================] - 804s - train loss: 1.0811     
Testing...
10000/10000 [==============================] - 47s - test loss: 0.9260   
```

the loss is decreasing at each epoch, but slower than advertised.

**Comment by rodrigob** [comment]
Bad news ! If I swtich `device=gpu0` to `device=cpu` then I get

```
rodrigob@lap-12-31 ~/c/r/k/examples> python3 mnist_cnn.py 
X_train shape: (60000, 1, 28, 28)
60000 train samples
10000 test samples
Train on 60000 samples, validate on 10000 samples
Epoch 0
60000/60000 [==============================] - 456s - loss: 0.4997 - acc: 0.8434 - val_loss: 0.1167 - val_acc: 0.9639
Epoch 1
60000/60000 [==============================] - 469s - loss: 0.1457 - acc: 0.9567 - val_loss: 0.0718 - val_acc: 0.9769
Epoch 2
60000/60000 [==============================] - 452s - loss: 0.1044 - acc: 0.9681 - val_loss: 0.0489 - val_acc: 0.9846
Epoch 3
60000/60000 [==============================] - 448s - loss: 0.0840 - acc: 0.9745 - val_loss: 0.0508 - val_acc: 0.9829
Epoch 4
60000/60000 [==============================] - 456s - loss: 0.0709 - acc: 0.9785 - val_loss: 0.0617 - val_acc: 0.9799
Epoch 5
60000/60000 [==============================] - 442s - loss: 0.0651 - acc: 0.9804 - val_loss: 0.0336 - val_acc: 0.9886
Epoch 6
60000/60000 [==============================] - 442s - loss: 0.0561 - acc: 0.9833 - val_loss: 0.0337 - val_acc: 0.9889
Epoch 7
60000/60000 [==============================] - 440s - loss: 0.0538 - acc: 0.9836 - val_loss: 0.0312 - val_acc: 0.9888
Epoch 8
60000/60000 [==============================] - 431s - loss: 0.0486 - acc: 0.9848 - val_loss: 0.0350 - val_acc: 0.9880
```

which means, the training works fine on cpu, but not working on gpu. This smells like a Theano bug (using latest git version); this is my first trial at using Theano... :(

**Comment by EderSantana** [comment]
do you have cudnn installed?

**Comment by fchollet** [comment]
This is probably the same issue as: https://github.com/fchollet/keras/issues/511

It seems this got resolved for everybody simply by upgrading to bleeding edge Theano and Keras. Try it...

**Comment by rodrigob** [comment]
Just tried that (one more time), still hitting issue #511 (nan loss)

**Comment by fchollet** [comment]
@rodrigob NaN loss would be a different issue, I believe. You can attempt to modify the epsilon values used in `optimizers` and `objectives` and see if larger epsilons solve the issue.

**Comment by rodrigob** [comment]
Tried today with fresh Keras and fresh Theano (again), and this time `python3 mnist_cnn.py` and `python3 cifar10_cnn.py`.  Hourray !

**Comment by log0** [comment]
@fchollet 

Actually not solved, I have a repro with the latest December builds. Could you please see where could this have gone wrong?

I am trying with Python 2 and Python 3, both repros with GPU on mnist_cnn.py, does not repro with CPU on mnist_cnn.py.

Theano - '0.7.0.dev-0db3f9afb20e6d73422474f0daa203bacbae1a88'
Keras - 0.3.0, built from github commit a18755dc67523afd1009093dcfa971d290bb20ab
GPU - GeForce GTX 780 

###
.theanorc

[global]
floatX = float32
device = gpu
optimizer = fast_compile
allow_gc = True

[lib]
cnmem = 0.6

[nvcc]
fastmath = True

I tried turning allow_gc to True/False, modifying cnmem from range 0.2 to 1.0, and modifying device from gpu/cpu. Only changing to cpu helps, but it takes about ~2500s per epoch then. Unusable speed.

###
Logs output:

$ python mnist_mlp.py 
Using Theano backend.
Using gpu device 0: GeForce GTX 780 (CNMeM is enabled)
60000 train samples
10000 test samples
Train on 60000 samples, validate on 10000 samples
Epoch 1/20
1s - loss: 0.2801 - acc: 0.9145 - val_loss: 0.1184 - val_acc: 0.9626
Epoch 2/20
1s - loss: 0.1150 - acc: 0.9654 - val_loss: 0.0976 - val_acc: 0.9680
Epoch 3/20
1s - loss: 0.0802 - acc: 0.9748 - val_loss: 0.0640 - val_acc: 0.9796
Epoch 4/20
1s - loss: 0.0626 - acc: 0.9805 - val_loss: 0.0736 - val_acc: 0.9771
Epoch 5/20
1s - loss: 0.0516 - acc: 0.9839 - val_loss: 0.0639 - val_acc: 0.9794
Epoch 6/20
1s - loss: 0.0408 - acc: 0.9867 - val_loss: 0.0634 - val_acc: 0.9799
Epoch 7/20
1s - loss: 0.0347 - acc: 0.9889 - val_loss: 0.0687 - val_acc: 0.9798
Epoch 8/20
1s - loss: 0.0300 - acc: 0.9901 - val_loss: 0.0690 - val_acc: 0.9800
Epoch 9/20
1s - loss: 0.0253 - acc: 0.9918 - val_loss: 0.0580 - val_acc: 0.9840
Epoch 10/20
1s - loss: 0.0222 - acc: 0.9927 - val_loss: 0.0556 - val_acc: 0.9848
Epoch 11/20
1s - loss: 0.0203 - acc: 0.9933 - val_loss: 0.0593 - val_acc: 0.9841
Epoch 12/20
1s - loss: 0.0179 - acc: 0.9939 - val_loss: 0.0618 - val_acc: 0.9850
Epoch 13/20
1s - loss: 0.0160 - acc: 0.9947 - val_loss: 0.0677 - val_acc: 0.9831
Epoch 14/20
1s - loss: 0.0135 - acc: 0.9954 - val_loss: 0.0677 - val_acc: 0.9837
Epoch 15/20
1s - loss: 0.0135 - acc: 0.9956 - val_loss: 0.0604 - val_acc: 0.9855
Epoch 16/20
1s - loss: 0.0113 - acc: 0.9961 - val_loss: 0.0586 - val_acc: 0.9862
Epoch 17/20
1s - loss: 0.0091 - acc: 0.9971 - val_loss: 0.0621 - val_acc: 0.9862
Epoch 18/20
1s - loss: 0.0095 - acc: 0.9969 - val_loss: 0.0675 - val_acc: 0.9833
Epoch 19/20
1s - loss: nan - acc: 0.3081 - val_loss: nan - val_acc: 0.0980
Epoch 20/20
1s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Test score: nan
Test accuracy: 0.098

**Comment by log0** [comment]
I also tried adding to theanorc "optimizer_excluding = cudnn", it helps but then it takes about ~2000s per epoch, again unusable.

**Comment by log0** [comment]
@fchollet

Actually not solved, I have a repro with the latest December builds. Could you please see where could this have gone wrong?

I am trying with Python 2 and Python 3, both repros with GPU on mnist_cnn.py, does not repro with CPU on mnist_cnn.py.

Theano - '0.7.0.dev-0db3f9afb20e6d73422474f0daa203bacbae1a88'
Keras - 0.3.0, built from github commit a18755d
GPU - GeForce GTX 780

###
.theanorc

[global]
floatX = float32
device = gpu
optimizer = fast_compile
allow_gc = True

[lib]
cnmem = 0.6

[nvcc]
fastmath = True

I tried turning allow_gc to True/False, modifying cnmem from range 0.2 to 1.0, and modifying device from gpu/cpu. Only changing to cpu helps, but it takes about ~2500s per epoch then. Unusable speed.

###
Logs output:

$ python mnist_cnn.py
Using Theano backend.
Using gpu device 0: GeForce GTX 780 (CNMeM is enabled)
X_train shape: (60000, 1, 28, 28)
60000 train samples
10000 test samples
Train on 60000 samples, validate on 10000 samples
Epoch 1/12
60000/60000 [==============================] - 7s - loss: nan - acc: 0.0988 - val_loss: nan - val_acc: 0.0980
Epoch 2/12
60000/60000 [==============================] - 7s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 3/12
60000/60000 [==============================] - 7s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 4/12
60000/60000 [==============================] - 7s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 5/12
60000/60000 [==============================] - 7s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 6/12
60000/60000 [==============================] - 7s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 7/12
60000/60000 [==============================] - 7s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 8/12
60000/60000 [==============================] - 7s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 9/12
60000/60000 [==============================] - 7s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 10/12
60000/60000 [==============================] - 7s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 11/12
60000/60000 [==============================] - 7s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 12/12
60000/60000 [==============================] - 7s - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Test score: nan
Test accuracy: 0.098

###

As you can see, the accuracy is very low and it's not really learning.

**Comment by c0d3rman** [comment]
Slightly unrelated: if `mnist_cnn.py` takes forever to run on your machine, install/fix cudnn. Worked for me.
Also example works out of the box for me after setting up the environment (with latest Keras).
